compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1934,9 +1936,146 @@ public BType getSafeType(BType type, boolean liftError) {
         return memberTypes;
     }
 
+    public boolean hasImplicitInitialValue(BType type) {
+        switch (type.tag) {
+            case TypeTags.INT:
+            case TypeTags.BYTE:
+            case TypeTags.FLOAT:
+            case TypeTags.DECIMAL:
+            case TypeTags.STRING:
+            case TypeTags.BOOLEAN:
+            case TypeTags.NIL:
+            case TypeTags.TYPEDESC:
+            case TypeTags.JSON:
+            case TypeTags.MAP:
+            case TypeTags.NONE:
+            case TypeTags.STREAM:
+            case TypeTags.TABLE:
+            case TypeTags.XML:
+            case TypeTags.ANYDATA:
+            case TypeTags.ANY:
+                return true;
+            case TypeTags.ANNOTATION:
+            case TypeTags.CHANNEL:
+            case TypeTags.ERROR:
+            case TypeTags.FUTURE:
+            case TypeTags.INVOKABLE:
+            case TypeTags.PACKAGE:
+            case TypeTags.SEMANTIC_ERROR:
+            case TypeTags.XML_ATTRIBUTES:
```

Comment:
don't need this cases right? can replace these from the `analyzeImplicitInitialValue` method's body

Reply:
I think having these tags are important for documenting that these types doesn't have IIV, what do you think? Because without performing any analysis we know that these types doesn't have IIVs.

Reply:
I don't think we need to add unnecessary code just for the purpose of documenting :)
'default' case means 'everything else'.

Reply:
For me this this reads like, 1st block does have IIVs for sure, 2nd block doesn't have for sure, for rest of it lets analyze the type to find out.

Reply:
in the type tags, if you re-order and move recod to just after the map, Your code will looks like below:
```ballerina
private boolean hasImplicitInitialValue(BType type) {
        if (type.tag < TypeTags.RECORD) {
            return true;
        }
        switch (type.tag) {
            case TypeTags.ARRAY:
                BArrayType arrayType = (BArrayType) type;
                if (arrayType.state == BArrayState.UNSEALED) {
                    // Empty array is the implicit initial value for arrays.
                    return true;
                }
                return hasImplicitInitialValue(arrayType.eType);
            case TypeTags.FINITE:
                return analyzeFiniteType((BFiniteType) type);
            case TypeTags.OBJECT:
                return analyzeObjectType((BObjectType) type);
            case TypeTags.RECORD:
                BRecordType recordType = (BRecordType) type;
                return recordType.fields.stream().allMatch(f -> hasImplicitInitialValue(f.type));
            case TypeTags.TUPLE:
                BTupleType tupleType = (BTupleType) type;
                return tupleType.tupleTypes.stream().allMatch(this::hasImplicitInitialValue);
            case TypeTags.UNION:
                return analyzeUnionType((BUnionType) type);
            default:
                return false;
        }
    }
``` 
Thats 50% reduction of LOC.
It still has the same three blocks you mentioned, but with positions swapped.

Reply:
Re-ordering RECORD breaks the compiler, it looks like there are some other place that depends on TAG ordering.
We can go with:
```java
        if (type.tag < TypeTags.RECORD) {
            return true;
        }
        switch (type.tag) {
            case TypeTags.TYPEDESC:
            case TypeTags.STREAM:
            case TypeTags.MAP:
                return true;
            case TypeTags.ARRAY:
                BArrayType arrayType = (BArrayType) type;
                if (arrayType.state == BArrayState.UNSEALED) {
                    // Empty array is the implicit initial value for arrays.
                    return true;
                }
                return hasImplicitInitialValue(arrayType.eType);
            case TypeTags.FINITE:
                return analyzeFiniteType((BFiniteType) type);
            case TypeTags.OBJECT:
                return analyzeObjectType((BObjectType) type);
            case TypeTags.RECORD:
                BRecordType recordType = (BRecordType) type;
                return recordType.fields.stream().allMatch(f -> hasImplicitInitialValue(f.type));
            case TypeTags.TUPLE:
                BTupleType tupleType = (BTupleType) type;
                return tupleType.tupleTypes.stream().allMatch(this::hasImplicitInitialValue);
            case TypeTags.UNION:
                return analyzeUnionType((BUnionType) type);
            default:
                return false;
        }
```